### PR TITLE
[Enhancement] Optimize mem usage of partial update

### LINKED ISSUE #11269: [Enhancement] Optimize the mem usage of partial update
## Enhancement

```
   @          0x2117d79 starrocks::faststring::GrowArray()
    @          0x1cd0479 starrocks::BinaryDictPageBuilder::BinaryDictPageBuilder()
    @          0x1c944bc starrocks::TypeEncodingTraits<>::create_page_builder()
    @          0x1c869f9 std::_Function_handler<>::_M_invoke()
    @          0x1c7fb12 starrocks::ScalarColumnWriter::set_encoding()
    @          0x1c80df5 starrocks::StringColumnWriter::speculate_column_and_set_encoding()
    @          0x1c7e0ef starrocks::StringColumnWriter::append()
    @          0x1b09049 starrocks::SegmentWriter::append_chunk()
    @          0x1cd9e48 starrocks::SegmentRewriter::rewrite()
    @          0x1c4ff0e starrocks::RowsetUpdateState::apply()
    @          0x1a9ca14 starrocks::TabletUpdates::_apply_rowset_commit()
    @          0x1aa0c23 starrocks::TabletUpdates::do_apply()
    @          0x2134b7d starrocks::ThreadPool::dispatch_thread()
    @          0x213038a starrocks::Thread::supervise_thread()
    @     0x7f4f98dc0dd5 start_thread
    @     0x7f4f983dbf6d __clone
```

Use column group read to reduce the mem usage of load large wide table for primary key model

### PR DESCRIPTION
**Pull request by sevev**
## What type of PR is this：
- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Which issues of this PR fixes ：

Fixes https://github.com/StarRocks/starrocks/issues/11269

## Problem Summary(Required) ：

We have partially optimized the primary key model for large import memory usage in this pr(https://github.com/StarRocks/starrocks/pull/12068), but the enhancement doesn't work if the load is partial update. And we also need a lot of memory if you do a large number of partial updates in one transaction. So this pr will try to reduce the memory usage of large partial update. 

There are two reasons for large memory usage during partial column updates:
1. The first one is that updating a few columns may increase the segment file size and we need to load all data of segment into memory which will cost a lot of memory.
2. The second one is that doing partial update requires reading data from other columns into memory, which can take up a lot of memory if the table has many columns.

In order to reduce memory usage,  the following two adjustments are made:
1. The first one is to estimate the length of the updated partial columns in each row when importing data, thus reducing the size of the segment file
2. The second one is don't load all the data of the rowset into memory at once, but to load them one by one according to the segment.

In my test env, one BE with two HDD, using StreamLoad, create a table with 65 column, 20 buckets:
```
CREATE TABLE `partial_test` (
  `col_1` bigint(20) NOT NULL COMMENT "",
  `col_2` bigint(20) NOT NULL COMMENT "",
  `col_3` bigint(20) NOT NULL COMMENT "",
  `col_4` varchar(150) NOT NULL COMMENT "",
  `col_5` varchar(150) NOT NULL COMMENT "",
  `col_6` varchar(150) NULL COMMENT "",
  `col_7` varchar(150) NULL COMMENT "",
  `col_8` varchar(1024) NULL COMMENT "",
  `col_9` varchar(120) NULL COMMENT "",
  `col_10` varchar(60) NULL COMMENT "",
  `col_11` varchar(10) NULL COMMENT "",
  `col_12` varchar(120) NULL COMMENT "",
  `col_13` varchar(524) NULL COMMENT "",
  `col_14` varchar(100) NULL COMMENT "",
  `col_15` varchar(150) NULL COMMENT "",
  `col_16` varchar(150) NULL COMMENT "",
  `col_17` varchar(150) NULL COMMENT "",
  `col_18` bigint(20) NULL COMMENT "",
  `col_19` varchar(500) NULL COMMENT "",
  `col_20` varchar(150) NULL COMMENT "",
  `col_21` tinyint(4) NULL COMMENT "",
  `col_22` int(11) NULL COMMENT "",
  `col_23` varchar(524) NULL COMMENT "",
  `col_24` bigint(20) NULL COMMENT "",
  `col_25` bigint(20) NULL COMMENT "",
  `col_26` varchar(8) NULL COMMENT "",
  `col_27` decimal64(18, 6) NULL COMMENT "",
  `col_28` decimal64(18, 6) NULL COMMENT "",
  `col_29` decimal64(18, 6) NULL COMMENT "",
  `col_30` decimal64(18, 6) NULL COMMENT "",
  `col_31` decimal64(18, 6) NULL COMMENT "",
  `col_32` decimal64(18, 6) NULL COMMENT "",
  `col_33` bigint(20) NULL COMMENT "",
  `col_34` decimal64(18, 6) NULL COMMENT "",
  `col_35` varchar(8) NULL COMMENT "",
  `col_36` decimal64(18, 6) NULL COMMENT "",
  `col_37` decimal64(18, 6) NULL COMMENT "",
  `col_38` varchar(8) NULL COMMENT "",
  `col_39` decimal64(18, 6) NULL COMMENT "",
  `col_40` decimal64(18, 6) NULL COMMENT "",
  `col_41` varchar(8) NULL COMMENT "",
  `col_42` decimal64(18, 6) NULL COMMENT "",
  `col_43` decimal64(18, 6) NULL COMMENT "",
  `col_44` decimal64(18, 6) NULL COMMENT "",
  `col_45` decimal64(18, 6) NULL COMMENT "",
  `col_46` int(11) NULL COMMENT "",
  `col_47` int(11) NOT NULL COMMENT "",
  `col_48` tinyint(4) NULL COMMENT "",
  `col_49` varchar(200) NULL COMMENT "",
  `col_50` tinyint(4) NULL COMMENT "",
  `col_51` varchar(200) NULL COMMENT "",
  `col_52` varchar(10) NULL COMMENT "",
  `col_53` tinyint(4) NULL COMMENT "",
  `col_54` tinyint(4) NULL COMMENT "",
  `col_55` varchar(150) NULL COMMENT "",
  `col_56` varchar(150) NULL COMMENT "",
  `col_57` varchar(500) NULL COMMENT "",
  `col_58` tinyint(4) NULL COMMENT "",
  `col_59` varchar(100) NULL COMMENT "",
  `col_60` varchar(150) NULL COMMENT "",
  `col_61` varchar(150) NULL COMMENT "",
  `col_62` varchar(150) NULL COMMENT "",
  `col_63` varchar(150) NULL COMMENT "",
  `col_64` datetime NULL COMMENT "",
  `col_65` datetime NULL COMMENT ""
) ENGINE=OLAP 
PRIMARY KEY(`col_1`, `col_2`, `col_3`)
COMMENT "OLAP"
DISTRIBUTED BY HASH(`col_1`, `col_2`) BUCKETS 20 
PROPERTIES (
"replication_num" = "1",
"in_memory" = "false",
"storage_format" = "V2",
"enable_persistent_index" = "true",
"compression" = "LZ4"
); 
```

|PrimaryKey Length| RowNum|BucketNum| Column Num| Partial ColumnNum | PartialUpdate RowsNum| Load time(s)| Apply time(ms)| Peak UpdateMemory usage | Note |
|---------------------|----------|------------|----------------|--------------------|------------------------------|----|-----|-----|----|
|12 Bytes| 300M | 20 |  65 | 5 | 100M | 135261 | 106693 | 78.9G | branch-main |
|12 Bytes| 300M | 20 |  65 | 5 | 100M | 166449| 149870 | 10.3G | branch-opt |
|12 Bytes| 300M | 20 |  65 | 5 | 100K | 2078 | 529 | 60.1M | branch-main |
|12 Bytes| 300M | 20 |  65 | 5 | 100K | 2211 | 541 | 60.2M | branch-opt |

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto backported to target branch
  - [x] 2.5
  - [x] 2.4
  - [x] 2.3
  - [ ] 2.2

### CODE REVIEW

**Comment by sevev** [comment]
run starrocks_admit_test

**Comment by sevev** [comment]
run starrocks_be_unittest

**Comment by sevev** [comment]
run starrocks_be_unittest

**Review by decster — DISMISSED** [review]

File: be/src/storage/delta_writer.cpp [hunk 1]
```diff
@@ -185,6 +185,13 @@ Status DeltaWriter::_init() {
             }
             writer_context.referenced_column_ids.push_back(index);
         }
+        int64_t average_row_size = _tablet->updates()->get_average_row_size();
+        if (average_row_size != 0) {
+            _memtable_buffer_row = config::write_buffer_size / average_row_size;
+        } else {
```
[the remark this answers is not in the document]
Comment (chaoyli) [reply be/src/storage/delta_writer.cpp:191]: You can use the default value to calculate the length.

File: be/src/storage/tablet_updates.h [hunk 2]
```diff
@@ -222,9 +225,16 @@ class TabletUpdates {
                              std::map<uint32_t, std::vector<uint32_t>>& rowids_by_rssid,
                              vector<std::unique_ptr<vectorized::Column>>* columns);
 
+    /*
```
Comment (chaoyli) [inline be/src/storage/tablet_updates.h:228]: The code annotated can be removed.

File: be/src/storage/delta_writer.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (sevev) [reply be/src/storage/delta_writer.cpp:203]: > When will the average_row_size be zero? If the average_row_size is zero, _memtable_buffer_row = 10240, may be a little small

If the tablet is a newly created tablet and has no historical data, the average_row_size is zero.

File: be/src/storage/memtable.h [hunk 3]
```diff
@@ -51,6 +52,10 @@ class MemTable {
 
     bool is_full() const;
 
+    void set_write_buffer_size(size_t max_buffer_size) { _max_buffer_size = max_buffer_size; }
```
[the remark this answers is not in the document]
Comment (sevev) [reply be/src/storage/memtable.h:65]: > Where does this function is called?

It is useless and has been removed.

File: be/src/storage/tablet_updates.h (hunk 2 above)
Comment (sevev) [reply be/src/storage/tablet_updates.h:241]: > The code annotated can be removed.

done

**Comment by sevev** [comment]
@chaoyli Your comments are resolved, PTAL again.

**Comment by sevev** [comment]
run starrocks_be_unittest

**Comment by sevev** [comment]
run starrocks_be_unittest

**Review by chaoyli — APPROVED** [review]

**Comment by sevev** [comment]
run starrocks_admit_test

**Comment by sevev** [comment]
run starrocks_admit_test

**Review by decster — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Comment by sevev** [comment]
@mergify backport branch-2.5

**Pull request merged**